File: src/tcore.erl
%% @doc Turnusol framework core module
%%
%% It does hot code swap with sync, GPIO pin initialization, GPIO pin write and read
%%
%% @end
%% @copyright 2014 Mahmut Bulut

-module (tcore).

-export([init_input_pin/1,
			init_output_pin/1,
			write_high/1,
			write_low/1,
			read_pin/1]).

-import (gpio, [write/2,
				read/1]).

% i2c, spi modules, usb?
% vsn ? version

-import (gpio_sup, [start_link/1]).

%%hello_world() -> io:fwrite("hello, world\n").

init_input_pin(Pin) -> start_link([{Pin, input}]).

init_output_pin(Pin) -> start_link([{Pin, output}]).

write_high(Pin) -> write(Pin, 1).

write_low(Pin) -> write(Pin, 0).

read_pin(Pin) -> read(Pin).

File: src/exsbase/tcore.ex
defmodule TCore do
  @moduledoc """
  Turnusol core functions
  """

  @doc """
  See Erlang implementation
  init_input_pin(Pin) -> start_link([{Pin, input}]).
  """
  def init_input_pin(pin) do
    :gpio_sup.start_link([{pin, :input}])
  end

  @doc """
  See Erlang implementation
  init_output_pin(Pin) -> start_link([{Pin, output}]).
  """
  def init_output_pin(pin) do
    :gpio_sup.start_link([{pin, :output}])
  end
  
  @doc """
  See Erlang implementation
  write_high(Pin) -> write(Pin, 1).
  """
  def write_high(pin) do
    :gpio.write(pin, 1)
  end

  @doc """
  See Erlang implementation
  write_low(Pin) -> write(Pin, 0).
  """
  def write_low(pin) do
    :gpio.write(pin, 0)
  end
  
  @doc """
  See Erlang implementation
  read_pin(Pin) -> read(Pin).
  """
  def read_pin(pin) do
    :gpio.read(pin)
  end

  @doc """
  Quits from shell and kills all nodes underlying the system
  """
  def kill_all do
    :erlang.halt()
  end
end